File: packages/server/src/entities/msg/msg.ts
import { Option } from "fp-ts/lib/Option";
import { AtRightError } from "../../at-error";
import { IAuthToken } from "../../auth";
import { IGenerator } from "../../generator";
import { Copyable } from "../../utils";
import { User } from "../user";
import { pipe } from "fp-ts/lib/pipeable";
import { option } from "fp-ts";

export interface IMsgAPI {
  readonly id: string;
  readonly priv: boolean;
  readonly text: string;
  readonly date: string;
}

export class Msg extends Copyable<Msg> {
  static create(
    objidGenerator: IGenerator<string>,
    receiver: Option<User>,
    text: string,
    now: Date,
  ): Msg {
    return new Msg(
      objidGenerator(),
      pipe(
        receiver,
        option.map(x => x.id),
      ),
      text,
      now,
    );
  }

  constructor(
    readonly id: string,
    readonly receiver: Option<string>,
    readonly text: string,
    readonly date: Date,
  ) {
    super(Msg);
  }

  toAPI(authToken: IAuthToken): IMsgAPI {
    if (
      pipe(
        this.receiver,
        option.map(x => x !== authToken.user),
        option.getOrElse(() => false),
      )
    ) {
      throw new AtRightError("アクセス権がありません。");
    }

    return {
      id: this.id,
      priv: pipe(
        this.receiver,
        option.isSome,
      ),
      text: this.text,
      date: this.date.toISOString(),
    };
  }
}

File: server-scala/src/main/scala/com/anontown/entities/msg/Msg.scala
package com.anontown.entities.msg

import java.time.OffsetDateTime
import cats._, cats.implicits._, cats.derived._
import com.anontown.utils.Impl._;
import com.anontown.AtError
import com.anontown.ports.ObjectIdGeneratorAlg
import com.anontown.ports.ClockAlg
import com.anontown.AtRightError
import com.anontown.AuthToken
import com.anontown.entities.user.{UserId, User}
import cats.data.EitherT

final case class MsgAPI(id: String, priv: Boolean, text: String, date: String);

object MsgAPI {
  implicit val implEq: Eq[MsgAPI] = {
    import auto.eq._
    semi.eq
  }
}

final case class Msg(
    id: MsgId,
    receiver: Option[UserId],
    text: String,
    date: OffsetDateTime
);

object Msg {
  implicit val implEq: Eq[Msg] = {
    import auto.eq._
    semi.eq
  }

  def create[F[_]: Monad: ObjectIdGeneratorAlg: ClockAlg](
      receiver: Option[User],
      text: String
  ): EitherT[F, AtError, Msg] = {
    for {
      id <- EitherT.right(ObjectIdGeneratorAlg[F].generateObjectId())
      requestDate <- EitherT.right(ClockAlg[F].getRequestDate())
    } yield Msg(
      id = MsgId(id),
      receiver = receiver.map(_.id),
      text = text,
      date = requestDate
    )
  }

  implicit class HistoryService(val self: Msg) {
    def toAPI(authToken: AuthToken): Either[AtError, MsgAPI] = {
      if (self.receiver.map(_ =!= authToken.user).getOrElse(false)) {
        Left(new AtRightError("アクセス権がありません。"))
      } else {
        Right(
          MsgAPI(
            id = self.id.value,
            priv = self.receiver.isDefined,
            text = self.text,
            date = self.date.toString()
          )
        )
      }
    }
  }
}
